### ISSUE TITLE
Add Attachments to the challenge specification

### ISSUE THREAD

**Issue by maxceem** [topcoder-platform/work-manager#917]
As API already supports it, we have to support adding attachments for the challenge specification.

FYI @vikasrohit

**Comment by rootelement** [comment]
Challenge API supports Attachments. You'll need to upload through the UI via the Filestack widget into the S3 bucket, then submit that s3 URL as the attachment URL. Then the challenge api will handle security and retrieval of the attachment.

**Comment by maxceem** [comment]
@vikasrohit file uploading is done via PR https://github.com/topcoder-platform/work-manager/pull/931

See demo video: https://monosnap.com/file/I4B4UsRVCtJBQv1wzraQwwvkcvibNF

Though there is some issue with config, so I use the incorrect FileStack/AWS bucket config, because it doesn't work for me with the correct one. As was clarified by @rootelement (on Slack) we have to upload challenge attachments to `tc-challenge-v5-dev` bucket. But if I config this bucket I get 403 error during uploading files to AWS via FileStack:

![image](https://user-images.githubusercontent.com/146016/101104631-38f74480-35d4-11eb-95c6-0da0b63b6c59.png)

But if I only change the bucket to `submission-staging-dev` the same code works for me, and I can upload files to AWS via FileStack:

![image](https://user-images.githubusercontent.com/146016/101104724-6fcd5a80-35d4-11eb-956b-54aeb6e3be5d.png)

But Challenge API doesn't allow creating attachments with `submission-staging-dev` because this bucket is not whitelisted. So to be able to test I had to set the bucket `tc-challenge-v5-dev` for attachments when creating them in Challenge API:

![image](https://user-images.githubusercontent.com/146016/101104919-e5392b00-35d4-11eb-86aa-96acf278706f.png)

So as result I upload files to `submission-staging-dev` but tell Challenge API that file is uploaded to `tc-challenge-v5-dev` so it accepts my POST requests 🤯

This might lead to the next issues I have:
- When I reload the challenge page, not all uploaded attachments are returned by Chalelnge API (even though all POST request were successful)
- I cannot download attachments as the server returns `Internal server error`. This is logical as I misconfigured buckets.
- I cannot remove attachments as the server returns `Internal server error`. This could be also logical if the server tries to clean file and cannot find it on the misconfigured bucket.

So we have to find out why we cannot upload files to `tc-challenge-v5-dev` AWS, while can upload to `submission-staging-dev`. There must be something with FileStack or AWS configuration.

**Comment by vikasrohit** [comment]
@rootelement @mtwomey can you guys help us here?

**Comment by rootelement** [comment]
I've reached out to Guna to fix the s3 bucket. @vikasrohit this is exactly like how the file upload in community app works. The idea is using the Filestack Dev key to upload to the `tc-challenge-v5-dev` bucket, then the challenge api should be able to read from it. I'm unclear how the frontend of filestack works, but if you're using this filestack api key in dev: `AzFINuQoqTmqw0QEoaw9az` you should be able to write to the `tc-challenge-v5-dev` bucket.

Matt and I will look into uploading to a workflow so things are virus scanned automatically.

I will confirm here when Guna has fixed the security issue.

**Comment by vikasrohit** [comment]
@rootelement Do we have confirmation for S3 bucket? @maxceem can you please give it a try in case it has been fixed?

**Comment by maxceem** [comment]
@vikasrohit it's still forbidden for `tc-challenge-v5-dev` bucket:

![image](https://user-images.githubusercontent.com/146016/101456292-82c98d00-393c-11eb-94da-2b5a1439034e.png)

**Comment by maxceem** [comment]
@vikasrohit Uploading attachments work well now end-to-end now.

There is only one issue with Challenge API when adding multiple attachments at once:
- we send multiple `POST /challenges/:id/attachments` requests in parallel and all of them are successful
- but after I reload the challenge `GET /challenge/:id` returns only one attachment

See demo video https://monosnap.com/file/mVqe0igPbDIa68bUepKwRujaTU0mQN

Example challenge `29d3d66f-d87f-402f-882f-258235ac7a6f` on DEV.

**Comment by vikasrohit** [comment]
Can we merge it now @maxceem ?

**Comment by vikasrohit** [comment]
@maxceem I am merging PR now as it seems the issue topcoder-platform/challenge-api#348  is fixed.

**Comment by maxceem** [comment]
@vikasrohit looks like something is not configured properly on DEV. When I'm trying to add attachments I'm getting errors from AWS services I guess:

![image](https://user-images.githubusercontent.com/146016/104346504-2c85d600-5508-11eb-8fbf-63fda46f4d1d.png)

**Comment by maxceem** [comment]
@vikasrohit locally the issue above is not reproducible and works good for me:

![image](https://user-images.githubusercontent.com/146016/104357407-d15ae000-5515-11eb-8935-3dd45fb0d64e.png)

The endpoint in challenge API has changed, so we would have to update UI code first before attachments creation would work.

So I've reverted the PR https://github.com/topcoder-platform/work-manager/pull/1035 for now, just in case we couldn't update it before the Jan 15th release.

**Comment by maxceem** [comment]
@vikasrohit code for attachments has been updated and pushed to DEV.

The only thing left is to configure `FILE_PICKER_API_KEY` environment variable on DEV and redeploy.

**Comment by rootelement** [comment]
I added the Filestack Dev API key to that key value in the dev appvar.json. Please redeploy and confirm.

**Comment by maxceem** [comment]
@rootelement I've rebuilt but looks like the env variable `FILE_PICKER_API_KEY` was not picked up, it still fallback to the dummy value. When building locally by `NODE_ENV=development FILE_PICKER_API_KEY=<KEY> npm run build` I can see that token is included in the built file.

![image](https://user-images.githubusercontent.com/146016/104655912-14a88080-56c7-11eb-826d-fa7e12153cb4.png)

**Comment by maxceem** [comment]
I've rechecked one more time.

If I run the next command `NODE_ENV=development npm run build ` without `FILE_PICKER_API_KEY` in the built file I can see a dummy value:

![image](https://user-images.githubusercontent.com/146016/104746016-66005080-5757-11eb-9936-53554a8b19cf.png)

But if I run `NODE_ENV=development FILE_PICKER_API_KEY=REAL_KEY_VALUE npm run build` in the built file I can see the value:

![image](https://user-images.githubusercontent.com/146016/104746359-d5764000-5757-11eb-8be3-5106187491cd.png)

And in the Work Manager built I can see the first situation as if the environment variable wasn't available during `npm rum build` script:

![image](https://user-images.githubusercontent.com/146016/104746441-f50d6880-5757-11eb-8b65-73a9d61f9194.png)

**Comment by rootelement** [comment]
Are we sure any of those environment variables are working in non-local deployments?

I know react is weird about reading env vars… From what I understand, you need to prepend REACT_APP_  on those values:
https://create-react-app.dev/docs/adding-custom-environment-variables/

Can you try that? I'll add the same value to the environment called `"REACT_APP_FILE_PICKER_API_KEY"` and we can test. The values there, can you try it and redeploy?

**Comment by maxceem** [comment]
Yeah, these variables should work on the server too. To clarify this situation, I've added the console.log of the env variable `FILE_PICKER_API_KEY` to the build script. The build script is just a regular script that is run on the server during the building process and as any regular server script, it should have access to the environment variables.

So when I run locally `NODE_ENV=development npm run build` I see:

![image](https://user-images.githubusercontent.com/146016/104807365-36008e00-57e7-11eb-945b-4117f0b5e1c4.png)

When I run `NODE_ENV=development FILE_PICKER_API_KEY=REAL_KEY_VALUE npm run build`:

![image](https://user-images.githubusercontent.com/146016/104807381-52042f80-57e7-11eb-8ebc-fb35f3f54bfb.png)

And when we run build in CricleCI:

![image](https://user-images.githubusercontent.com/146016/104807648-179b9200-57e9-11eb-8705-5ee934b20c59.png)

**Comment by maxceem** [comment]
### QA guidelines:

- Upload a single attachment
- Upload multiple attachments
- Upload attachments with a not supported extension like `.none` - should get error
- Upload multiple attachments: some with correct extension and some with forbidden extension
- Should not allow uploading more than 10 attachments
- Upload by drag & drop
- Upload by clicking using the upload dialog
- Delete attachments
- Download attachments from EDIT page 
- Download attachments from VIEW page
- As we don't show attachments in Community APP so far what we can do, is:
   - Check that we can download attachments form VIEW page
   - Open Browser Dev Tools 
   - Activate the challenge with attachments
   - Find active challenge it in the challenge list https://www.topcoder-dev.com/challenges?tracks[DS]=true&tracks[Des]=true&tracks[Dev]=true&tracks[QA]=true&types[]=CH&types[]=F2F&types[]=TSK and click it to open the challenge with attachment. 
   - Then find challenge request in the network tab, the challenge object should contain array with attachments:
   
      ![image](https://user-images.githubusercontent.com/146016/105008598-1f09a800-5a42-11eb-91a5-1b06fde52a8f.png)

**Comment by drasticdpk** [comment]
FYI 
As discussed with @vikasrohit , I am making it fail. we will test it later.

**Comment by maxceem** [comment]
Uploading attachments on DEV works good. But on PROD there seems to be some config issue as uploading fails during uploading files to AWS S3 due to `403` permission errors:

start was good:
![image](https://user-images.githubusercontent.com/146016/105148672-52fcd000-5b0b-11eb-8f98-05e33ec16381.png)

upload was good:
![image](https://user-images.githubusercontent.com/146016/105148699-598b4780-5b0b-11eb-9e57-9f2bf53629d2.png)

but actual uploading to S3 failed because of authentication issue `403`:
![image](https://user-images.githubusercontent.com/146016/105148733-6314af80-5b0b-11eb-84eb-f518c7b510ea.png)

**Comment by vikasrohit** [comment]
Moving it out to next milestone after Project Defaults shape up project as we are not able to get it working in production. We need latest challenge api in production to get it working.

**Comment by acshields** [comment]
@vikasrohit - noted.

**Comment by maxceem** [comment]
@vikasrohit looks like Challenge API has been deployed to PROD https://github.com/topcoder-platform/challenge-api/compare/master...develop so we might give it another try.

**Comment by vikasrohit** [comment]
Yep, we can. But I would like to work on it only when we have its consumers i.e. some app is using the attachments.

**Comment by vikasrohit** [comment]
@acshields @rootelement  do you know if Community/Forums app is already have code or plans to use the challenge Attachments?

**Comment by rootelement** [comment]
I believe Community App will read attachments. I think that needs to be tested. Forums uses their own attachments mechanism, but i don't believe they're reading challenge-api attachments at all. The security on the endpoint for attachments requires you're registered on the challenge as a submitter (or above... copilot, manager, observer) to get the attachments.

**Comment by vikasrohit** [comment]
Okay, thanks for the updates @rootelement. @acshields let us know how you want to prioritize it, IMO, we should do it once we have confirmation on the usage of attachments.

**Comment by acshields** [comment]
@vikasrohit - for my clarity, when you say confirmation on the usage of attachments, are you referring to the forums usage of attachments?

**Comment by vikasrohit** [comment]
I meant any kind of usage of the challenge attachments. I am not sure where exactly we want to consume the attachments of the challenge, but what I want to say is that we should retest this feature and deploy to prod when we know there is some consumer of this feature whether it is Community app or forums.

**Comment by vikasrohit** [comment]
@acshields moving it out from bug fix cycle.

**Comment by acshields** [comment]
@vikasrohit - noted. Right now, I have this as an aspect of the new UI implementation for WM. If that project pitch gets accepted for the next round, we could group this into that effort.

**Comment by yoution** [comment]
@phead198708 please assign to me

**Comment by yoution** [comment]
@vikasrohit  @acshields can you provide the new UI implementation?
currently the ui is like so 
![image](https://user-images.githubusercontent.com/3863914/124244463-8bd8b080-db51-11eb-9385-4cd1c9e51a0a.png)

**Comment by vikasrohit** [comment]
Sorry @yoution I guess this issue is not ready yet.
